### PR TITLE
basic support of NRPN, ref #154

### PR DESCRIPTION
**Pull request by yaxu**

### CODE REVIEW

File: classes/DirtEventTypes.sc [hunk 1]
```diff
@@ -38,18 +38,18 @@ DirtEventTypes {
...
 					if(~ccn.notNil) { midicmd = \control; ~ctlNum = ~ccn };
 					if(~ccv.notNil) { midicmd = \control; ~control = ~ccv };
 					if(~nrpn.notNil) {
 						midicmd = \control;
-						~nrpnLSB = ~nrpn % 128;
-                        ~nrpnMSB = (~nrpn - ~nrpnLSB) / 128;
-						~valLSB = ~val % 128;
-                        ~valMSB = (~val - ~valLSB) / 128;
+						nrpnLSB = ~nrpn % 128;
+                        nrpnMSB = (~nrpn - nrpnLSB) / 128;
```
Comment (telephon) [inline classes/DirtEventTypes.sc]: could you check the formatting? Mabe you need to use be tabs, not spaces.

File: classes/DirtEventTypes.sc [hunk 2]
```diff
@@ -81,24 +80,23 @@ DirtEventTypes {
...
-							midiout.control(~chan, 99, ~nrpnMSB);
-							midiout.control(~chan, 98, ~nrpnLSB);
-							midiout.control(~chan, 6,  ~valMSB);
-							midiout.control(~chan, 38, ~valLSB)
+						if (nrpnMSB.notNil) {
+							midiout.control(chan, 99, nrpnMSB);
+							midiout.control(chan, 98, nrpnLSB);
+							midiout.control(chan, 6,  valMSB);
+							midiout.control(chan, 38, valLSB)
 						}
```
Comment (telephon) [inline classes/DirtEventTypes.sc]: usually, we use the syntax
```
if(something) {

} {

}
```

File: classes/DirtEventTypes.sc [hunk 3]
```diff
@@ -81,24 +80,23 @@ DirtEventTypes {
...
 					{
-						if (~nrpn.notNil) {
+						if (nrpnMSB.notNil) {
 							thisThread.clock.sched(latency, {
-								midiout.control(0,99,0);
-								midiout.control(~chan, 99, ~nrpnMSB);
-								midiout.control(~chan, 98, ~nrpnLSB);
-								midiout.control(~chan, 6,  ~valMSB);
-								midiout.control(~chan, 38, ~valLSB)
+								midiout.control(chan, 99, nrpnMSB);
```
Comment (telephon) [inline classes/DirtEventTypes.sc]: you don't need to duplicate this code. Jus define a function further up and call it in both cases. Functions are first-class, so they are just something like:

```
var sendNRPN = {
	midiout.control(chan, 99, nrpnMSB);
	midiout.control(~chan, 98, ~nrpnLSB);
	midiout.control(chan, 98, nrpnLSB);
	midiout.control(~chan, 6,  ~valMSB);
	midiout.control(chan, 6,  valMSB);
	midiout.control(~chan, 38, ~valLSB)
	midiout.control(chan, 38, valLSB)
};

```

**Review by telephon — CHANGES_REQUESTED** [review]

File: classes/DirtEventTypes.sc (hunk 1 above)
Comment (yaxu) [reply classes/DirtEventTypes.sc]: Is it any better now?

File: classes/DirtEventTypes.sc (hunk 3 above)
Comment (yaxu) [reply classes/DirtEventTypes.sc]: Thanks, done!

File: classes/DirtEventTypes.sc (hunk 1 above)
Comment (telephon) [reply classes/DirtEventTypes.sc]: no, not yet! Not sure what it is. You can select and hit tab in the IDE for autoformat. And show invisible characters to double check that there are tabs not spaces.

Comment (yaxu) [reply classes/DirtEventTypes.sc]: Ok sorry, sorted now. ("Show spaces and tabs" seems to have no effect on my 3.10.3 supercollider compiled under linux.)

**Review by telephon — APPROVED** [review]

**Pull request merged**